fix: add migration for profile syncing controller

Prithpal-Sooriya opened the pull request:
<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been completely filled out, and PR status checks have passed at least once.
-->

## **Description**

For existing users, we want to ensure that profile syncing is disabled (opt-in). This migration ensures that the `isProfileSyncingEnabled` controller state is changed to false for existing users.

This will need to be cherry-picked into v12.0.0

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/26004?quickstart=1)

## **Related issues**

Fixes: V12 testing issue of the state for profile syncing.

## **Manual testing steps**

1. Go to this page...
2.
3.

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

Prithpal-Sooriya committed:
fix: add migration for profile syncing controller

we want existing users to have profile syncing disabled, this will remove/nullify the isEnableProfileSyncing state field.

Prithpal-Sooriya commented on app/scripts/controllers/user-storage/user-storage-controller.ts:
```diff
@@ -27,15 +27,15 @@ export type UserStorageControllerState = {
    * Condition used by UI and to determine if we can use some of the User Storage methods.
    */
-  isProfileSyncingEnabled: boolean;
+  isProfileSyncingEnabled: boolean | null;
```
Checked all references (JS and TS files) it seems okay to change this field to be nullable.

Prithpal-Sooriya replied:
We do not perform strict boolean checks in references, it is truthy checks (`isProfileSyncingEnabled && ...` or `isProfileSyncingEnabled || ...`)

Prithpal-Sooriya committed:
Merge branch 'develop' of github.com:MetaMask/metamask-extension into fix/add-profile-syncing-enabled-migration

Prithpal-Sooriya committed:
test: add some additional tests for this migration

Gudahtt commented on app/scripts/controllers/user-storage/user-storage-controller.ts:
```diff
@@ -27,15 +27,15 @@ export type UserStorageControllerState = {
 
 const defaultState: UserStorageControllerState = {
-  isProfileSyncingEnabled: true,
+  isProfileSyncingEnabled: null,
```
Shouldn't this still be `true` for new users?

Prithpal-Sooriya replied:
WHOOPS. Yes this is correct, was too quick with it!

Gudahtt commented on app/scripts/migrations/120.1.ts:
```diff
@@ -0,0 +1,37 @@
+  return versionedData;
+}
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
```
We should use `unknown` here rather than `any`. See here for why we should be avoiding the use of `any` in general: https://github.com/MetaMask/contributor-docs/blob/main/docs/typescript.md#avoid-any

Prithpal-Sooriya replied:
Valid, this was copied from a previous migration. Migrated to using `unknown`

Gudahtt commented on app/scripts/migrations/120.1.ts:
```diff
@@ -0,0 +1,37 @@
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+function transformState(state: Record<string, any>) {
+  if (!hasProperty(state, 'UserStorageController')) {
```
Pre-existing users would be expected to have no `UserStorageController` state, since this is the first time this controller is introduced. Is that true? We should be expecting this condition to be `true` in all cases. This is the case where we want to set `isProfileSyncingEnabled` to `null`.

Prithpal-Sooriya replied:
This is interesting. The PR that it was introduced was 11.16 (but this label may be incorrect).
https://github.com/MetaMask/metamask-extension/pull/23353

I think you are right though, we can add a check for if it does exist too. I'll update the tests as well.

Gudahtt replied:
Oh interesting, so it was introduced before it was used, and not visible in settings during that time. OK, if that's true, then you're right. We'll need to handle both cases.

Prithpal-Sooriya committed:
fix: fix small mistake and also cover other edge cases during the migration

we need to also handle migration of new users who migrated from 11.16 that might have UserStorage in their state

Gudahtt commented on app/scripts/migrations/120.1.test.ts:
```diff
@@ -0,0 +1,54 @@
+    });
+  });
+
+  it('does nothing if user storage is not initialized', async () => {
```
We'll want to update this test description

Prithpal-Sooriya replied:
Good eye lol. Yep will update!

Prithpal-Sooriya committed:
test: update test names

Prithpal-Sooriya commented:
Tested the migration on v12 and v11. Seems to work as intended

Prithpal-Sooriya committed:
chore: update e2e test snapshots

Gudahtt commented on app/scripts/migrations/120.1.ts:
```diff
@@ -0,0 +1,48 @@
+    return state;
+  }
+
+  // Existing users who do have UserStorageController state.
+  // nullify `isProfileSyncingEnabled`
+  if (
+    isObject(state.UserStorageController) &&
+    hasProperty(state.UserStorageController, 'isProfileSyncingEnabled')
+  ) {
+    state.UserStorageController.isProfileSyncingEnabled = null;
+    return state;
+  }
```
Nit: It looks like we have an uncovered branch here, the case where the UserStorageController state is invalid. We should handle the case where the `UserStorageController` is set incorrectly (e.g. the state was corrupted).

Usually in cases like this, if it's totally unfathomable how it could be set to anything different, we would submit a Sentry error. Note that this would require importing `captureException` and adding a new test as well.

```suggestion
  if (
    !isObject(state.UserStorageController) ||
    !hasProperty(state.UserStorageController, 'isProfileSyncingEnabled')
  ) {
    captureException(
      `Migration ${version}: Invalid UserStorageController state: ${typeof state.PreferencesController}`,
    );
    return state;
  }

  // Existing users who do have UserStorageController state.
  // nullify `isProfileSyncingEnabled`
  state.UserStorageController.isProfileSyncingEnabled = null;
```

Prithpal-Sooriya replied:
Oooooh yep ty for raising this. Yeah I'll add some capture exceptions in case for corrupted state. Will add tests too!

Gudahtt reviewed (dismissed):
LGTM! One pending suggestion about how to handle corrupted state, which would be ideal to fix but doesn't seem critical enough to block.

Prithpal-Sooriya committed:
refactor: handle and capture migration failures due to malformed state

Gudahtt commented:
Hmm, still some e2e test failure in the error capture tests

Prithpal-Sooriya committed:
Merge branch 'develop' into fix/add-profile-syncing-enabled-migration

matteoscurati reviewed (approved):
Go!
